<img style="float: left; margin: 30px 15px 15px 15px;" src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTFzQj91sOlkeDFkg5HDbjtR4QJYmLXkfMNig&usqp=CAU" width="500" height="500" /> 
    
    
### <font color='navy'> Simulación matemática. 

**Nombres:** Óscar Guillermo Chiw Lozano.

**Fecha:** 29 de octubre del 2023.

**Expediente** : 736814.
    
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/Guillermo-Chiw-ITESO/G9_T7ChiwGuillermo

# Tarea 8: Clase 18

## Tarea.
> - Consultar tasas de interés reales en algún banco y proyectar un ahorro mensual para que al terminar su carrera tengan $\$85000$ MXN en su cuenta.

In [2]:
# Librería de análisis de datos
import pandas as pd
# Librería para manipulación de fechas
from datetime import date
import numpy as np
import numpy_financial as npf

### Ejercicio 1: Tasa de interés de GBM (10.4%) para 14 meses.

Código de solución estudiante 1

In [3]:
# Datos para la cuenta de ahorro
tasa_interes = 0.104
years = 1.17
Abono_year = 12
inicial = 25000
meta = 85000
fecha_inicio = (date(2023,11,1))

In [4]:
# Cálculo de abonos mensuales iguales para llegar a la meta de ahorro en el tiempo deseado
Abono = -npf.pmt(tasa_interes/Abono_year,years*Abono_year,(meta-inicial))
Abono

4557.244956735618

In [5]:
# Rango de fechas en la tabla
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'
# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index = rng,columns=['Abono','Balance'],dtype='float')
# Para poner el indice con los periodos y no con las fechas
df.reset_index(inplace=True)

df.index+=1
df.index.name='Periodo'
df.head()

Fecha del abono  Abono  Balance
Periodo                                
1            2023-11-30    NaN      NaN
2            2023-12-31    NaN      NaN
3            2024-01-31    NaN      NaN
4            2024-02-29    NaN      NaN
5            2024-03-31    NaN      NaN

In [6]:
npf.ppmt(tasa_interes/Abono_year, 1, years * Abono_year, Abono)

-306.64523696982843

In [7]:
# Los abonos son iguales
df['Abono'] = Abono
# Vector indizador de los periodos de los abonos
index_vector = np.arange(1,len(rng)+1)
# Extracto de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1)/(tasa_interes/Abono_year))

# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)
df

Fecha del abono    Abono   Balance
Periodo                                   
1            2023-11-30  4557.24  29773.91
2            2023-12-31  4557.24  34589.20
3            2024-01-31  4557.24  39446.22
4            2024-02-29  4557.24  44345.33
5            2024-03-31  4557.24  49286.90
6            2024-04-30  4557.24  54271.30
7            2024-05-31  4557.24  59298.89
8            2024-06-30  4557.24  64370.06
9            2024-07-31  4557.24  69485.18
10           2024-08-31  4557.24  74644.63
11           2024-09-30  4557.24  79848.80
12           2024-10-31  4557.24  85098.06
13           2024-11-30  4557.24  90392.82
14           2024-12-31  4557.24  95733.47

### Ejercicio 2: Tasa de interés de BBVA (3.25%) para 14 meses.

In [8]:
# Datos para la cuenta de ahorro
tasa_interes = 0.0325
years = 1.17
Abono_year = 12
inicial = 25000
meta = 85000
fecha_inicio = (date(2023,11,1))

# Cálculo de abonos mensuales iguales para llegar a la meta de ahorro en el tiempo deseado
Abono = -npf.pmt(tasa_interes/Abono_year,years*Abono_year,(meta-inicial))
Abono

4361.052915351751

In [9]:
# Rango de fechas en la tabla
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'
# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index = rng,columns=['Abono','Balance'],dtype='float')
# Para poner el indice con los periodos y no con las fechas
df.reset_index(inplace=True)

df.index+=1
df.index.name='Periodo'
df.head()

Fecha del abono  Abono  Balance
Periodo                                
1            2023-11-30    NaN      NaN
2            2023-12-31    NaN      NaN
3            2024-01-31    NaN      NaN
4            2024-02-29    NaN      NaN
5            2024-03-31    NaN      NaN

In [10]:
npf.ppmt(tasa_interes/Abono_year, 1, years * Abono_year, Abono)

-305.1685238625558

In [11]:
# Los abonos son iguales
df['Abono'] = Abono
# Vector indizador de los periodos de los abonos
index_vector = np.arange(1,len(rng)+1)
# Extracto de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1)/(tasa_interes/Abono_year))

# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)
df

Fecha del abono    Abono   Balance
Periodo                                   
1            2023-11-30  4361.05  29428.76
2            2023-12-31  4361.05  33869.52
3            2024-01-31  4361.05  38322.30
4            2024-02-29  4361.05  42787.14
5            2024-03-31  4361.05  47264.08
6            2024-04-30  4361.05  51753.14
7            2024-05-31  4361.05  56254.35
8            2024-06-30  4361.05  60767.76
9            2024-07-31  4361.05  65293.40
10           2024-08-31  4361.05  69831.28
11           2024-09-30  4361.05  74381.46
12           2024-10-31  4361.05  78943.97
13           2024-11-30  4361.05  83518.83
14           2024-12-31  4361.05  88106.08

### Ejercicio 3: Tasa de interés de Hey Banco (12.00%) para 14 meses.

In [16]:
# Datos para la cuenta de ahorro
tasa_interes = 0.12
years = 1.17
Abono_year = 12
inicial = 25000
meta = 85000
fecha_inicio = (date(2023,11,1))

# Cálculo de abonos mensuales iguales para llegar a la meta de ahorro en el tiempo deseado
Abono = -npf.pmt(tasa_interes/Abono_year,years*Abono_year,(meta-inicial))
Abono

4601.819220466113

In [17]:
# Rango de fechas en la tabla
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'
# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index = rng,columns=['Abono','Balance'],dtype='float')
# Para poner el indice con los periodos y no con las fechas
df.reset_index(inplace=True)

df.index+=1
df.index.name='Periodo'
df.head()

Fecha del abono  Abono  Balance
Periodo                                
1            2023-11-30    NaN      NaN
2            2023-12-31    NaN      NaN
3            2024-01-31    NaN      NaN
4            2024-02-29    NaN      NaN
5            2024-03-31    NaN      NaN

In [18]:
npf.ppmt(tasa_interes/Abono_year, 1, years * Abono_year, Abono)

-306.92747675952785

In [19]:
# Los abonos son iguales
df['Abono'] = Abono
# Vector indizador de los periodos de los abonos
index_vector = np.arange(1,len(rng)+1)
# Extracto de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1)/(tasa_interes/Abono_year))

# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)
df

Fecha del abono    Abono   Balance
Periodo                                   
1            2023-11-30  4601.82  29851.82
2            2023-12-31  4601.82  34752.16
3            2024-01-31  4601.82  39701.50
4            2024-02-29  4601.82  44700.33
5            2024-03-31  4601.82  49749.15
6            2024-04-30  4601.82  54848.46
7            2024-05-31  4601.82  59998.77
8            2024-06-30  4601.82  65200.58
9            2024-07-31  4601.82  70454.40
10           2024-08-31  4601.82  75760.76
11           2024-09-30  4601.82  81120.19
12           2024-10-31  4601.82  86533.21
13           2024-11-30  4601.82  92000.36
14           2024-12-31  4601.82  97522.19

### Respuesta planteada en el ejercicio usando (markdown)
 #### Banco GBC (tasa de interés del 10.4%):
| Fecha del abono |  Abono  | Balance   |
|-----------------|---------|-----------|
| Periodo         |         |           |
| 1               | 2023-11-30 | 4557.24 | 29773.91 |
| 2               | 2023-12-31 | 4557.24 | 34589.20 |
| 3               | 2024-01-31 | 4557.24 | 39446.22 |
| 4               | 2024-02-29 | 4557.24 | 44345.33 |
| 5               | 2024-03-31 | 4557.24 | 49286.90 |
| 6               | 2024-04-30 | 4557.24 | 54271.30 |
| 7               | 2024-05-31 | 4557.24 | 59298.89 |
| 8               | 2024-06-30 | 4557.24 | 64370.06 |
| 9               | 2024-07-31 | 4557.24 | 69485.18 |
| 10              | 2024-08-31 | 4557.24 | 74644.63 |
| 11              | 2024-09-30 | 4557.24 | 79848.80 |
| 12              | 2024-10-31 | 4557.24 | 85098.06 |
| 13              | 2024-11-30 | 4557.24 | 90392.82 |
| 14              | 2024-12-31 | 4557.24 | 95733.47 |

#### Banco BBVA (tasa de interés del 3.25%):
| Fecha del abono |  Abono  | Balance   |
|-----------------|---------|-----------|
| Periodo         |         |           |
| 1               | 2023-11-30 | 4361.05 | 29428.76 |
| 2               | 2023-12-31 | 4361.05 | 33869.52 |
| 3               | 2024-01-31 | 4361.05 | 38322.30 |
| 4               | 2024-02-29 | 4361.05 | 42787.14 |
| 5               | 2024-03-31 | 4361.05 | 47264.08 |
| 6               | 2024-04-30 | 4361.05 | 51753.14 |
| 7               | 2024-05-31 | 4361.05 | 56254.35 |
| 8               | 2024-06-30 | 4361.05 | 60767.76 |
| 9               | 2024-07-31 | 4361.05 | 65293.40 |
| 10              | 2024-08-31 | 4361.05 | 69831.28 |
| 11              | 2024-09-30 | 4361.05 | 74381.46 |
| 12              | 2024-10-31 | 4361.05 | 78943.97 |
| 13              | 2024-11-30 | 4361.05 | 83518.83 |
| 14              | 2024-12-31 | 4361.05 | 88106.08 |

#### Banco Hey Banco (tasa de interés del 12.0%):
| Fecha del abono |  Abono  | Balance   |
|-----------------|---------|-----------|
| Periodo         |         |           |
| 1               | 2023-11-30 | 4601.82 | 29851.82 |
| 2               | 2023-12-31 | 4601.82 | 34752.16 |
| 3               | 2024-01-31 | 4601.82 | 39701.50 |
| 4               | 2024-02-29 | 4601.82 | 44700.33 |
| 5               | 2024-03-31 | 4601.82 | 49749.15 |
| 6               | 2024-04-30 | 4601.82 | 54848.46 |
| 7               | 2024-05-31 | 4601.82 | 59998.77 |
| 8               | 2024-06-30 | 4601.82 | 65200.58 |
| 9               | 2024-07-31 | 4601.82 | 70454.40 |
| 10              | 2024-08-31 | 4601.82 | 75760.76 |
| 11              | 2024-09-30 | 4601.82 | 81120.19 |
| 12              | 2024-10-31 | 4601.82 | 86533.21 |
| 13              | 2024-11-30 | 4601.82 | 92000.36 |
| 14              | 2024-12-31 | 4601.82 | 97522.19 |
